[#36204] Add new GitHub plugin tab enabling users to copy git actions

tessi opened the pull request:
### Features

* It registers a new tab to the new core work package tab system.
* The GitHub tab is designed according to the figma prototype
* It allows copying default branch names, commit messages, and git-feature-branch-creation shell commands using [`tickety-tick-formatter`](https://github.com/bitcrowd/tickety-tick-formatter) under the hood which we just extracted from `tickety-tick` for this purpose.
* Two new icons are introduced to our icon font, a lighter console icon and a git-merge icon
* It is intentionally not yet displaying pull requests (this part will be continued in a separate PR)
* the ruby-side of the GitHub integration was refactored slightly to be without rubocop warnings (there are no intentional functional changes)

https://user-images.githubusercontent.com/206108/108431746-94d08180-7243-11eb-83b0-61bf13a4c6a6.mov

### State of this PR 🏗️ 

This PR is complete feature-wise but still has a couple of TODOs:

* [ ] get an angular-architecture review from @oliverguenther 
* [x] write js tests
* [x] write feature specs
* [x] get #8999 merged to unblock merging this PR
* [x] fix linter errors
* [x] https://github.com/bitcrowd/tickety-tick-formatter was created as a package for this PR. We should probably also review that to some degree
* [x] CLA

tessi committed:
Let GitHub integration show changes in a separate tab

* added new github integration icons to icon font
* add tab content:  working tab-header and copy-menu

tessi commented on frontend/src/app/components/op-context-menu/op-context-menu-handler.ts:
```diff
@@ -18,11 +18,11 @@ export abstract class OpContextMenuHandler extends UntilDestroyedMixin {
 
   public onOpen(menu:JQuery) {
-    menu.find('.menu-item').first().focus();
+    menu.find('.menu-item').first().trigger('focus');
```
this is a drive-by fix - I noticed that `focus()` is a deprecated method while reading some menu handler code. The deprecation notice says to use `trigger()` instead.

b12f commented on modules/github_integration/frontend/module/tab-header/tab-header.template.html:
```diff
@@ -0,0 +1,30 @@
+    <li class="toolbar-item">
+      <button class="button github-create-pr"
+          [disabled]="true"
+          [attr.aria-label]="text.createPrButtonDescription">
```
It's usually good form to make buttons `type="button"`. They're `type="submit"` by default, and if this tab header gets included into a larger form switching tabs might submit the form by accident.

tessi replied:
I made all buttons `type="button"`, thanks for the recommendation :)

tessi commented:
Update to the video in the PR description: The copy button is now placed next to the textarea, but looks attached especially in the one-line version.


https://user-images.githubusercontent.com/206108/108552745-01a25500-72f2-11eb-9395-902d44bdaa2d.mov

tessi commented on frontend/src/app/components/datetime/timezone.service.spec.ts:
```diff
@@ -59,7 +59,7 @@ describe('TimezoneService', function () {
     });
 
-    timezoneService = TestBed.get(TimezoneService);
+    timezoneService = TestBed.inject(TimezoneService);
```
Drive-by fix 🚜: this (and following similar changes) are fixing a deprecation warning I spotted while reading code

tessi commented on frontend/src/app/modules/hal/resources/work-package-resource.spec.ts:
```diff
@@ -48,8 +48,6 @@ import {WorkPackageNotificationService} from 'core-app/modules/work_packages/not
 import {TimezoneService} from 'core-components/datetime/timezone.service';
 import {APIV3Service} from 'core-app/modules/apiv3/api-v3.service';
-import { Tab } from 'core-app/components/wp-single-view-tabs/tab/tab';
-import { TabComponent } from 'core-app/components/wp-single-view-tabs/tab/tab.component';
```
These were just unused and introduced in my former PR. I wonder why the linter didn't complain 🤔  Anyways, now they're gone

lindenthal commented:
Hi @tessi,
can you please provide the [CLA](https://docs.openproject.org/development/#openproject-contributor-license-agreement-cla) for your contribution?
Thanks
Niels

tessi commented:
Note: The "Create PR" Button was removed again as we decided this feature does not makes sense at the moment

tessi commented:
> Hi @tessi,
> can you please provide the [CLA](https://docs.openproject.org/development/#openproject-contributor-license-agreement-cla) for your contribution?
> Thanks
> Niels

Note: We talked about this in a private conversation - from my perspective the CLA discussion is resolved ✅

Commit:
modernise github_integrations ruby code

* refactored some code to be more modern ruby (if wrote most of it 7
  years ago and couldn't look at some parts without squinting too much)
* make some intended-to-be-private module methods actually private
* fixed all rubocop errors in the /modules/github_integration
* re-organized tests a little
* gave our rubocop.yml some RSpec-related defaults -- happy to discuss
  these, but I think we can live with these as a good starting point

👆 all without actually (intentionally) changing the behaviour

tessi commented on .rubocop.yml:
```diff
@@ -7,6 +7,12 @@ AllCops:
+
+Gemspec/RequiredRubyVersion:
+  Exclude:
+    - modules/**/*.gemspec
```
IMO: inline module "gems" should not define their own ruby version and instead rely on the main definition

ulferts replied:
:+1:

tessi commented on .rubocop.yml:
```diff
@@ -282,5 +288,12 @@ Style/HashTransformKeys:
 
-Rspec/MultipleMemoizedHelpers:
+
+RSpec/MultipleMemoizedHelpers:
+  Max: 15
+
+RSpec/MultipleExpectations:
   Max: 15
+
+RSpec/ExampleLength:
+  Max: 25
```
In my attempt to bring you a github_integration without rubocop warnings, I stumbled over some RSpec-related rules which didn't have good defaults yet (in my opinion, of course).

I'd suggest these as relaxed good defaults as a starting point. We can probably adjust these to our needs over time.

ulferts replied:
:+1:

Commit:
removed dead angular template code

Commit:
codeclimate found more things than rubocop :)

Commit:
removed create-pr-button since we decided against implementing that feature

aleixsuau reviewed (commented):
Hi @tessi ,

The frontend part LGTM. Regarding the Ruby part, I think it would be better if someone else more suited than me could take a look.

Thanks!

tessi commented on frontend/src/global_styles/fonts/_openproject_icon_definitions.scss:
```diff
@@ -1,1536 +1,1548 @@
 @mixin icon-mixin-accessibility {
-    content: "\f101";
+    content: '\f101';
```
code climate wanted me to change just two of these definitions but had all these quotes in their internal list of "known bad formatting errors". So I fixed them all instead of just my two quotes. 🤷

ulferts commented on modules/github_integration/lib/open_project/github_integration/patches/api/work_package_representer.rb:
```diff
@@ -0,0 +1,24 @@
+        def extension
+          ->(*) do
+            link :github,
+                 uncacheable: true do
```
The part of caching in OP relevant here works by having most of the JSON response cached and then removing parts from it. As long as the part is valid for all users, as is the case here, we only need to make sure that those users not allowed to see it don't. For that, OP offers `cache_if`. For an example see the [ProjectRepresenter](https://github.com/opf/openproject/blob/dev/lib/api/v3/projects/project_representer.rb#L71).

You can thus simply move the condition into the `cache_if` lambda.

ulferts replied:
On second thoughts, it should be enough to reduce the check to only:

```
cache_if: -> { current_user.allowed_to?(:show_github_content, represented.project) }
```

The permission will only be granted if the project supports it, i.e. if the module is enabled.

tessi replied:
cool, wouldn't have guess that from the name `cache_if` (thought it sometimes caches and sometimes not). But great to know, makes the code much shorter

ulferts commented on modules/github_integration/config/locales/js-en.yml:
```diff
@@ -0,0 +1,54 @@
+          copy_success: ✅ Copied!
+          copy_error: ❌ Copy failed!
+      tab_prs:
+        empty: There are no Pull Requests linked yet. Link an existing PR by using the code <code>OP#%{wp_id}</code> in the PR description or create a new PR.
```
I assume "Pull Request" is not a name so it would need to be written with lower case -> "pull request"

tessi replied:
true! I somehow thought they made it a trademark/name. but apparently not. fixed!

ulferts commented on modules/github_integration/lib/open_project/github_integration/engine.rb:
```diff
@@ -52,5 +54,16 @@ class Engine < ::Rails::Engine
+      OpenProject::AccessControl.map do |ac_map|
+        ac_map.project_module(:github, {}) do |pm_map|
+          pm_map.permission(:show_github_content, {}, {})
+        end
```
an I18n key for `project_module_github` is missing which is why it falls back to 'Github' instead of displaying 'GitHub':

![image](https://user-images.githubusercontent.com/617519/113720782-455cec80-96ef-11eb-8f2f-f6941613a689.png)

ulferts replied:
Same is true for `permission_show_github_content`

![image](https://user-images.githubusercontent.com/617519/113721337-bb615380-96ef-11eb-9f23-4959d5e7eb1b.png)

ulferts reviewed (changes requested):
@tessi, I only found some minor things. So should be good to merge soon.

Commit:
added missing translations

Commit:
properly cache the github related part of the wp api

Commit:
lower case pull requests in translations

Commit:
fix specs

tessi commented:
@ulferts thanks for the review 💚 (TIL `cache_if`)  all points are addressed now
